text/0000-pattern-3.md

```diff
@@ -0,0 +1,1653 @@
+                | C | D | E | F | G |                        | E | F |
+               -+---+---+---+---+---+-                       +---+---+
+                        ^_______^                            ^       ^
+                         5..7                                5       7
+
+               -+---+---+---+---+---+-                               +---+
+                | C | D | E | F | G |                                | G |
+               -+---+---+---+---+---+-                               +---+
+                                ^___^                                ^   ^
+                                 7..8                                7   8
+```
+
+A span of shared haystack will always store a copy of the original haystack when splitting,
+because the haystack can be cheaply cloned. Splitting is thus just manipulation of the range only.
+Slicing is only done when returning from an algorithm.
+
+A non-shared haystack needs to maintain unique ownership for each haystack slice. Therefore,
+a haystack will be split as soon as the span is split. The "original range" becomes a value
+disconnected from the haystack, and this is where `.restore_range()` is needed:
+to recover the indices in the middle (`5 == 3 + 2` and `7 == 3 + 4`).
+
+### Searcher
+
+A searcher has two required methods `.search()` and `.consume()`,
+and an optional method `.trim_start()`.
+
+```rust
+pub unsafe trait Searcher<A: Hay + ?Sized> {
+    fn search(&mut self, span: Span<&A>) -> Option<Range<A::Index>>;
+    fn consume(&mut self, span: Span<&A>) -> Option<A::Index>;
+    fn trim_start(&mut self, hay: &A) -> A::Index { ... }
+}
+
+pub unsafe trait ReverseSearcher<A: Hay + ?Sized>: Searcher<A> {
+    fn rsearch(&mut self, span: Span<&A>) -> Option<Range<A::Index>>;
+    fn rconsume(&mut self, span: Span<&A>) -> Option<A::Index>;
+    fn trim_end(&mut self, hay: &A) -> A::Index { ... }
+}
+
+pub unsafe trait DoubleEndedSearcher<A: Hay + ?Sized>: ReverseSearcher<A> {}
+```
+
+`.search()` and `.consume()` are safe because there is no safe ways to construct a `Span<&A>`
+with invalid ranges. Implementations of these methods often start with:
+
+```rust
+    fn search(&mut self, span: SharedSpan<&A>) -> Option<Range<A::Index>> {
+        let (hay, range) = span.into_parts();
+        // search for pattern from `hay` restricted to `range`.
+    }
+```
+
+The trait is unsafe to implement because it needs to guarantee the returned range is valid.
+
+The `.search()` method will look for the first slice matching the searcher's pattern in the span,
+and returns the range where the slice is found (relative to the hay's start index).
+The `.consume()` method will is similar, but anchored to the start of the span.
+
+```rust
+let span = unsafe { Span::from_parts("CDEFG", 3..8) };
+// we can find "CD" at the start of the span.
+assert_eq!("CD".into_searcher().search(span.clone()), Some(3..5));
+assert_eq!("CD".into_searcher().consume(span.clone()), Some(5));
```

Comment:
Should this (and the other examples calling `.into_searcher().consume(...)`) be `.into_consumer().consume(...)`?

Reply:
Fixed. Copy-and-paste error.